packages/astro/src/core/csp/common.ts

```diff
@@ -8,24 +8,35 @@ import type { AstroConfig, CspAlgorithm } from '../../types/public/index.js';
 import type { BuildInternals } from '../build/internal.js';
 import { generateCspDigest } from '../encryption.js';
 
-export function shouldTrackCspHashes(config: AstroConfig): boolean {
-	return config.experimental?.csp === true || typeof config.experimental?.csp === 'object';
+export function shouldTrackCspHashes(
+	csp: any,
+): csp is Exclude<AstroConfig['experimental']['csp'], false> {
```

Comment:
nit: you could extract the type above in the file to reuse it in all functions eg